aioxmpp/muc/service.py

```diff
@@ -697,27 +825,147 @@ def _inbound_muc_user_presence(self, stanza):
                     stanza.from_,
                 )
                 return
-            self._occupant_info[info.occupantjid] = info
-            self.on_join(stanza, info)
+            self._occupant_info[info.conversation_jid] = info
+            self.on_join(info)
             return
 
         mode, data = self._diff_presence(stanza, info, existing)
         if mode == _OccupantDiffClass.NICK_CHANGED:
             new_nick, = data
-            del self._occupant_info[existing.occupantjid]
-            existing.occupantjid = existing.occupantjid.replace(
+            old_nick = existing.nick
+            del self._occupant_info[existing.conversation_jid]
+            existing._conversation_jid = existing.conversation_jid.replace(
                 resource=new_nick
             )
-            self._occupant_info[existing.occupantjid] = existing
-            self.on_nick_change(stanza, existing)
+            self._occupant_info[existing.conversation_jid] = existing
+            self.on_nick_changed(existing, old_nick, new_nick)
         elif mode == _OccupantDiffClass.LEFT:
             mode, actor, reason = data
             existing.update(info)
-            self.on_leave(stanza, existing, mode, actor=actor, reason=reason)
-            del self._occupant_info[existing.occupantjid]
+            self.on_leave(existing,
+                          muc_leave_mode=mode,
+                          muc_actor=actor,
+                          muc_reason=reason)
+            del self._occupant_info[existing.conversation_jid]
 
     @asyncio.coroutine
-    def change_nick(self, new_nick):
+    def send_message(self, msg):
+        """
+        Send a message to the MUC.
+
+        :param msg: The message to send.
+        :type msg: :class:`aioxmpp.Message`
+
+        There is no need to set the address attributes or the type of the
+        message correctly; those will be overridden by this method to conform
+        to the requirements of a message to the MUC. Other attributes are left
+        untouched (except that :meth:`~.StanzaBase.autoset_id` is called) and
+        can be used as desired for the message.
+
+        .. seealso::
+
+            :meth:`.AbstractConversation.send_message` for the full interface
+            specification.
+        """
+        msg.type_ = aioxmpp.MessageType.GROUPCHAT
+        msg.to = self._mucjid
+        msg.xep0045_muc_user = muc_xso.UserExt()
```

Comment:
I’m following suggestion 1 from <https://mail.jabber.org/pipermail/standards/2017-January/032048.html>:

> Current best practice is for MUC implementations to add a tag to all messages (groupchat and PMs) sent from the MUC:
>
>        <x xmlns="http://jabber.org/protocol/muc#user">
>
> Carbons implementations do check for presence of this tag and disable CC'ing accordingly. We need to codify this, and it belongs into both XEP-0045 (adding the <x/> tag) and into XEP-0280 (using the tag as a filter criterion). I'm willing to create PRs for both once consensus is reached ("45 version bump", anyone?).

I also realise just now that this doesn’t affect us, because we do not "send from the MUC" and are thus not the MUC implementation Georg is talking about there. However, it doesn’t hurt and may in fact help if the MUC we’re using doesn’t follow this suggestion and also does not strip the ``<x/>`` we’re adding.